staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,143 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+)
+
+// Decorator composes an inner http.ResponseWriter object, the decorator
+// object should return the inner (decorated) http.ResponseWriter
+// object via the Unwrap function.
+type Decorator interface {
+	http.ResponseWriter
+
+	// Unwrap returns the inner http.ResponseWriter object associated with the decorator
+	Unwrap() http.ResponseWriter
+}
+
+// Wrap accepts a Decorator object and wraps with an "outer" http.ResponseWriter
+// object that implements http.CloseNotifier, http.Flusher and http.Hijacker accordingly
+func Wrap(decorator Decorator) http.ResponseWriter {
+	inner := decorator.Unwrap()
+
+	_, notifiable := inner.(http.CloseNotifier)
+	_, hijackable := inner.(http.Hijacker)
+	_, flushable := inner.(http.Flusher)
+
+	// from go net/http documentation:
+	// The default HTTP/1.x and HTTP/2 ResponseWriter implementations support Flusher
+	// Handlers should always test for this ability at runtime.
+	//
+	// The Hijacker interface is implemented by ResponseWriters that allow an HTTP handler
+	// to take over the connection.
+	// The default ResponseWriter for HTTP/1.x connections supports Hijacker, but HTTP/2 connections
+	// intentionally do not. ResponseWriter wrappers may also not support Hijacker.
+	// Handlers should always test for this ability at runtime
+	//
+	// The CloseNotifier interface is implemented by ResponseWriters which allow detecting
+	// when the underlying connection has gone away.
+	// Deprecated: the CloseNotifier interface predates Go's context package.
+	// New code should use Request.Context instead.
+	switch {
+	case notifiable && hijackable && flushable:
+		// for HTTP/1.x request the default implementation of ResponseWriter
+		// also implement CloseNotifier, Flusher and Hijacker
+		return &withCloseNotifyFlushAndHijack{
+			withCloseNotifyAndFlush: &withCloseNotifyAndFlush{
+				Decorator: decorator,
+			},
+		}
+	case notifiable && flushable:
+		// for HTTP/2 request, the default ResponseWriter object (http2responseWriter)
+		// implements Flusher and CloseNotifier.
+		return &withCloseNotifyAndFlush{
+			Decorator: decorator,
+		}
+	default:
+		// we should never be here for either http/1.x or http2 request
+		return decorator
+	}
+}
+
+// GetOriginal goes through the chain of wrapped http.ResponseWriter objects
+// and returns the original http.ResponseWriter object provided by net/http
```

Comment:
"net/http" is a little too specific here, the chain of Wraps can be based on other sorts of ResponseWriters.

Reply:
updated the comment